src/main/java/org/reaktivity/nukleus/tls/internal/stream/TlsServerFactory.java

```diff
@@ -241,18 +236,69 @@ private MessageConsumer newNetworkStream(
         final long routeId = begin.routeId();
         final long authorization = begin.authorization();
 
-        final MessagePredicate filter = (t, b, o, l) -> true;
-        final RouteFW route = router.resolve(routeId, authorization, filter, wrapRoute);
+        final MessagePredicate emptyFilter = (t, b, o, l) -> true;
+        final RouteFW route = router.resolve(routeId, authorization, emptyFilter, wrapRoute);
 
         MessageConsumer newStream = null;
 
         if (route != null)
         {
+            final long resolvedId = route.correlationId();
             final long initialId = begin.streamId();
 
-            final TlsServer server = new TlsServer(network, routeId, initialId, authorization);
+            final TlsStoreInfo tlsStoreInfo = lookupStore.apply(RouteId.localId(resolvedId));
+            final SSLContext sslContext = tlsStoreInfo == null ? null : tlsStoreInfo.context;
+            final SSLEngine tlsEngine = sslContext == null ? null : sslContext.createSSLEngine();
 
-            newStream = server::onNetwork;
+            if (tlsEngine != null)
+            {
+                tlsEngine.setUseClientMode(false);
+
+                if (tlsStoreInfo.supportsClientAuth)
+                {
+                    tlsEngine.setWantClientAuth(true);
+                }
+
+                tlsEngine.setHandshakeApplicationProtocolSelector((ex, ps) ->
+                {
+                    SSLSession session = tlsEngine.getHandshakeSession();
+                    byte[] tlsHostnameEncoded = null;
+                    if (session instanceof ExtendedSSLSession)
+                    {
+                        ExtendedSSLSession sessionEx = (ExtendedSSLSession) session;
+                        List<SNIServerName> serverNames = sessionEx.getRequestedServerNames();
+                        if (!serverNames.isEmpty())
+                        {
+                            SNIServerName serverName = serverNames.get(0);
+                            tlsHostnameEncoded = serverName.getEncoded();
+                        }
+                    }
+
+                    if (tlsHostnameEncoded != null)
+                    {
+                        tlsHostnameRO.wrap(tlsHostnameEncoded, 0, tlsHostnameEncoded.length);
+                    }
+
+                    final MessagePredicate filter = (t, b, i, l) ->
+                    {
+                        final TlsRouteExFW routeExFW = wrapRouteEx.apply(t, b, i, l);
+
+                        final String8FW hostname = routeExFW.hostname();
+
+                        return hostname != null && Objects.equals(tlsHostnameRO, hostname.value());
+                    };
```

Comment:
```
                    final DirectBuffer tlsHostname0 = tlsHostname;
                    final MessagePredicate filter = (t, b, i, l) ->
                    {
                        final TlsRouteExFW routeExFW = wrapRouteEx.apply(t, b, i, l);

                        final String8FW hostname = routeExFW.hostname();

                        return hostname != null && Objects.equals(tlsHostname0, hostname.value());
                    };
```
We must make sure to use the right value for comparison, even if we don't call `tlsHostnameRO.wrap(...)`.